### PR TITLE
Detailed report for testing.assert_equal and testing.assert_identical

### PR DESCRIPTION
**Pull request by benbovy**
- ~~Closes #xxxx~~
 - [x] Tests added / passed
 - [x] Passes ``git diff upstream/master | flake8 --diff``
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

~~In addition to `Dataset` repr, the error message also shows the output of `Dataset.info()` for both datasets.~~

~~This may not be the most elegant solution, but it is helpful when datasets only differ by their attributes attached to coordinates or data variables (not shown in repr). I'm open to any suggestion.~~

The report shows the differences for dimensions, data values (``Variable`` and ``DataArray``), coordinates, data variables and attributes (the latter only for ``testing.assert_identical``).

There is currently not much tests for `xarray.testing` functions, but I'm willing to add more if needed.

Not sure if it's worth a what's new entry (EDIT: added one).

### CODE REVIEW

File: xarray/testing.py [hunk 1]
```diff
@@ -82,7 +90,10 @@ def assert_identical(a, b):
         assert a.name == b.name
         assert_identical(a._to_temp_dataset(), b._to_temp_dataset())
     elif isinstance(a, (xr.Dataset, xr.Variable)):
-        assert a.identical(b), '{}\n{}'.format(a, b)
+        assert a.identical(b), (
+            '{}\n{}\n---\n{}\n{}'
+            .format(a, b, _get_info_as_str(a), _get_info_as_str(b))
+        )
```
Comment (jhamman) [inline xarray/testing.py]: if this is a `xr.Variable` instance, this will not work since it won't have the info method.

**Comment by shoyer** [comment]
This is certainly welcome. In an ideal world, we might point specifically to the exact differing variables/attributes, but this is a fine start.

**Comment by benbovy** [comment]
Indeed, exact differing variables/attributes would be much better here. I'll look into this next week. 

[pytest's failure reports](https://docs.pytest.org/en/stable/example/reportingdemo.html) and reports of `numpy.testing` functions may be good sources of inspiration.

**Comment by max-sixty** [comment]
Any interest in pushing this over the line @benbovy ?

**Comment by benbovy** [comment]
Thanks for reminding me this PR @max-sixty ! It deserves more than no commits since 1.5 years!

The assertion error messages now look like [pytest's failure reports](https://docs.pytest.org/en/stable/example/reportingdemo.html) (see ``test_eq_dict``). I've tried to reuse as much as possible the repr formatting that we already have for the xarray objects.

As examples:

```python
ds_a = xr.Dataset(data_vars={'var1': (('x', 'y'), [[1, 2, 3], [4, 5, 7]]),
                             'var2': ('x', [3, 4])},
                  coords={'x': ['a', 'b'], 'y': [1, 2, 3]},
                  attrs={'units': 'm', 'description': 'desc'})

ds_b = xr.Dataset(data_vars={'var1': ('x', [1, 2])},
                  coords={'x': ('x', ['a', 'c'], {'source': 0}), 'label': ('x', [1, 2])},
                  attrs={'units': 'kg'})
```
```
>>> xr.testing.assert_identical(ds_a, ds_b)
AssertionError: Left and right Dataset objects are not identical
Differing dimensions:
(x: 2, y: 3) != (x: 2)
Differing coordinates:
L * x        (x) <U1 'a' 'b'
R * x        (x) <U1 'a' 'c'
    source: 0
Left contains more coordinates:
  * y        (y) int64 1 2 3
Right contains more coordinates:
    label    (x) int64 1 2
Differing data variables:
L   var1     (x, y) int64 1 2 3 4 5 7
R   var1     (x) int64 1 2
Left contains more data variables:
    var2     (x) int64 3 4
Differing attributes:
L   units: m
R   units: kg
Left contains more attributes:
    description: desc
```

```python
da_a = xr.DataArray([[1, 2, 3], [4, 5, 7]],
                    dims=('x', 'y'),
                    coords={'x': ['a', 'b'], 'y': [1, 2, 3]},
                    attrs={'units': 'm', 'description': 'desc'})

da_b = xr.DataArray([1, 2],
                    dims='x',
                    coords={'x': ['a', 'c'], 'label': ('x', [1, 2])},
                    attrs={'units': 'kg'})
```
```
>>> xr.testing.assert_equal(da_a, da_b)
AssertionError: Left and right DataArray objects are not equal
Differing values:
L
    array([[1, 2, 3],
           [4, 5, 7]])
R
    array([1, 2])
Differing coordinates:
L * x        (x) <U1 'a' 'b'
R * x        (x) <U1 'a' 'c'
Left contains more coordinates:
  * y        (y) int64 1 2 3
Right contains more coordinates:
    label    (x) int64 1 2
```

Those examples are probably not the most readable ones but it's for a full showcase.

A downside of this approach (i.e., report the exact differing) is that when ``a`` and ``b`` differs, full comparison is done once again for formatting the report. I don't think there is an easy way to avoid that, but I don't think it's a big deal either.

File: xarray/core/formatting.py [hunk 2]
```diff
@@ -463,3 +468,128 @@ def dataset_repr(ds):
...
+            diff_items += [ab_side + s[1:]
+                           for ab_side, s in zip(('L', 'R'), temp)]
+
+    if diff_items:
+        summary += ["Differing {}:".format(title)] + diff_items
+
+    summary += extra_items_repr(a_keys - b_keys, a_mapping, "Left")
+    summary += extra_items_repr(b_keys - a_keys, b_mapping, "Right")
+
+    return "\n".join(summary)
```
[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/core/formatting.py]: No need for Py2 compat!

[the remark this answers is not in the document]
Comment (benbovy) [reply xarray/core/formatting.py]: Great!

**Comment by max-sixty** [comment]
>  full comparison is done once again for formatting the report. I don't think there is an easy way to avoid that, but I don't think it's a big deal either.

i.e. there's a performance cost? Yeah that's def fine!

**Comment by max-sixty** [comment]
This looks amazing! I'm excited to use it. Thanks for finishing it up

**Comment by benbovy** [comment]
I'm wondering which is the best among the options below:

A. 

```
Differing data variables:
L   var1     (x, y) int64 1 2 3 4 5 7
R   var1     (x) int64 1 2
L   var2     (x, y) int64 4 5 6 7 8 9
    description: variable 2 
R   var2     (x, y) int64 4 5 6 7 8 9
```

B.

```
Differing data variables:
L   
    var1     (x, y) int64 1 2 3 4 5 7
R
    var1     (x) int64 1 2
L
    var2     (x, y) int64 4 5 6 7 8 9
    description: variable 2 
R
    var2     (x, y) int64 4 5 6 7 8 9
```

C.

```
Differing data variables:
L   
    var1     (x, y) int64 1 2 3 4 5 7
    var2     (x, y) int64 4 5 6 7 8 9
    description: variable 2
R
    var1     (x) int64 1 2
    var2     (x, y) int64 4 5 6 7 8 9
```

**Comment by max-sixty** [comment]
I think you should feel free to merge an MVP rather than perfecting, but check out https://docs.pytest.org/en/latest/example/simple.html#writing-well-integrated-assertion-helpers

I _think_ that might be as simple as wrapping these in `pytest.fail(...)`

**Comment by max-sixty** [comment]
> I'm wondering which is the best among the options below:

TBH I think they're all good; I don't have a strong view. I'd low-confidence rank A->C->B

**Comment by dcherian** [comment]
I agree with @max-sixty A > C > B but only weakly.

**Comment by max-sixty** [comment]
This looks v good @benbovy!

**Comment by benbovy** [comment]
> I agree with @max-sixty A > C > B but only weakly.

Same opinion.

> This looks v good @benbovy!

I'm just fixing `int32` vs `int64` issue with tests run on appveyor, then to me it's ready to merge unless someone else has objections.

File: xarray/tests/test_formatting.py [hunk 3]
```diff
@@ -190,6 +193,97 @@ def test_attribute_repr(self):
...
+        Differing values:
+        L
+            array([[1, 2, 3],
+                   [4, 5, 7]])
+        R
+            array([1, 2])
+        Differing coordinates:
+        L * x        (x) <U1 'a' 'b'
+        R * x        (x) <U1 'a' 'c'
+        Left contains more coordinates:
```
Comment (shoyer) [inline xarray/tests/test_formatting.py]: "Left contains more coordinates" sounds a little funny to me.

Maybe "Coordinates on the left DataArray but not the right" or "Coordinates only on the left object"?

Comment (benbovy) [reply xarray/tests/test_formatting.py]: Agreed! I followed pytest's reports too blindly here.

**Comment by benbovy** [comment]
If everyone is happy with this I'm going to merge it.

**Comment by shoyer** [comment]
> If everyone is happy with this I'm going to merge it.

Yes, please!

**Comment by max-sixty** [comment]
Thanks @benbovy!

**Pull request merged**